stdlib/CMakeLists.txt

```diff
@@ -38,6 +38,12 @@ else()
   set(SWIFT_STDLIB_STABLE_ABI_default FALSE)
 endif()
 
+if(SWIFT_BUILD_SDK_OVERLAY OR SWIFT_INCLUDE_TESTS)
```

Comment:
This code seems to be assuming that we never invoke CMake in this directory directly, we always go through the top-level `CMakeLists.txt` in the `swift/` directory. Is that assumption true (at least, it's not obvious to me that it's true)? Otherwise, these variables will be uninitialized.

We hit a similar issue in https://github.com/apple/swift/pull/34371#issue-507213229:

> In direct CMake calls to stdlib/public/Platform/CMakeLists.txt, the variable SWIFT_STDLIB_STABLE_ABI is not initialized.

Reply:
This is a very good point. I think the situation as it is today is like this: cmake/modules/StandaloneOverlay.cmake is a hack that's *trying* to avoid invoking the top-level CMakeLists.txt, but it's already full of problems and fragility. I certainly don't want to break it, but I don't think we should be actively adding support for this type of build either. In this particular case, I believe this means we're "okay" because cmake/modules/StandaloneOverlay.cmake isn't used to build any of the modules that are involved in the new SWIFT_BUILD_TEST_SUPPORT_MODULES flag. And even if they were, notice that the changes around SWIFT_BUILD_TEST_SUPPORT_MODULES I'm making are purely inclusive, i.e. when we build the modules in all the existing cases + also if SWIFT_BUILD_TEST_SUPPORT_MODULES is set.

So in terms of making sure we don't break cmake/modules/StandaloneOverlay.cmake I think we're good.

The more longer-term strategy, though, should *actually* be to make stdlib/CMakeLists.txt standalone and not depending on the top-level CMakeLists.txt. But AFAIK, today, stdlib/CMakeLists.txt doesn't work separately, yet. I think that definitions of CMake variables and options that are only affecting the runtime/stdlib and not tools (swiftc) should eventually be moved to stdlib/CMakeLists.txt or perhaps elsewhere inside stdlib/. But that doesn't belong in this PR — let me try to move the directly affected ones here (SWIFT_BUILD_DYNAMIC_SDK_OVERLAY, SWIFT_BUILD_SDK_OVERLAY) as a separate PR.